spacy_experimental/biaffine_parser/arc_predicter.pyx

```diff
@@ -323,21 +325,23 @@ def split_lazily(docs: List[Doc], *, ops: Ops, max_tokens: int, senter_name: str
             raise ValueError(f"Lazy splitting requires senter pipe `{senter_name}` to have ",
                               "`save_activations` enabled.\nDuring training, `senter` must be "
                               "in the list of annotating components.")
-        scores = activations['probabilities']
-        split_recursive(scores[:,1], ops, max_tokens, lens)
+        scores = activations['probabilities'][:, 1]
+        _split_lazily_doc(ops, scores, max_tokens, lens)
 
     assert sum(lens) == sum([len(doc) for doc in docs])
 
     return ops.asarray1i(lens)
 
 
-def split_recursive(scores, ops, max_tokens, lengths):
-    if len(scores) < max_tokens:
-        lengths.append(len(scores))
-    else:
-        # Find the best splitting point. Exclude the first token, because it
-        # wouldn't split the current partition (leading to infinite recursion).
-        start = ops.xp.argmax(scores[1:]) + 1
-
-        split_recursive(scores[:start], ops, max_tokens, lengths)
-        split_recursive(scores[start:], ops, max_tokens, lengths)
+def _split_lazily_doc(ops: Ops, scores: Floats2d,  max_tokens: int, lens: List[int]):
+    q = deque([scores])
+    while q:
+        scores = q.popleft()
+        if len(scores) <= max_tokens:
+            lens.append(len(scores))
+        else:
+            # Find the best splitting point. Exclude the first token, because it
+            # wouldn't split the current partition (leading to infinite recursion).
+            start = ops.xp.argmax(scores[1:]) + 1
+            q.appendleft(scores[start:])
+            q.appendleft(scores[:start])
```

Comment:
Why was the order swapped here?

Reply:
The queue is processed left-to-right (so in sequence order) and we want to process the left split first, so we have to push it last.

Reply:
I can rewrite it a bit as a typical stack to make it clearer, but the order will still be inverted.